### ISSUE TITLE
Go indexer: Handle unnamed struct field initializers. [phab:D1699]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2126]
_Created by **fromberger** at 2017-06-07 20:02:34_: 

When constructing a struct literal, paint references from each initializing
expression to the field it initializes.

Add tests for same.

**Comment by kythe** [comment]
_**jdennett** wrote at 2017-06-07 21:15:47_:

At the level of the schema I'd really like these to be distinguishable from named references to a field.  The distinction seems important (and is why historically we've chosen not to have these).

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-06-07 21:36:37_:

diagnostic-worthy?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-06-07 21:36:37_:

`ref/anon`?

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-07 21:38:37_:

>>! In (earlier comment), jdennett wrote:
> At the level of the schema I'd really like these to be distinguishable from named references to a field.  The distinction seems important (and is why historically we've chosen not to have these).

That's a very reasonable point. I thought about proposing a `ref/initializes` sort of edge kind, and we could certainly still do that if it matters.

My memory is probably imperfect, but I thought one of the main reasons we omitted these in the original design was the problem of overlapping and large-spanning links in the UI. Do you recall a more semantically-motivated case? I'd like to document that, except I don't recall. :)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-07 21:41:05_:

I don't really think so; the only cases where this should happen are when we have //already// had failures from the type-checker. I'm not yet emitting diagnostics for those, but the indexer has captured them and I'm planning to do so in a future pass. So I think it's not necessary to add one here, which is essentially a random usage site. What do you think?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-07 21:41:05_:

I think that's somewhat to the point of jdennett's question. Maybe `ref/initializes` rather than `anon`? 

zarko do you have opinions on this?

**Comment by kythe** [comment]
_**jdennett** wrote at 2017-06-08 01:07:00_:

`ref/anon` appeals to me, but that's just an opinion.  A use case: the field is renamed, and a client wants to determine which code to update for that.  ref/anon edges require no update.  Another: fields are reordered.  *Only* ref/anon initializers need to be updated.

Another way to look at this (and maybe name it) would be in terms of named field references vs positional field references; somewhat similar to languages with function parameters that can be initialized either by name or by position at the option of a caller.

(I don't see this as blocking, but it would be nice to figure this out for the schema and then make sure the code matches.)

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-08 02:34:53_:

>>! In (earlier comment), jdennett wrote:
> `ref/anon` appeals to me, but that's just an opinion.  A use case: the field is renamed, and a client wants to determine which code to update for that.  ref/anon edges require no update.  Another: fields are reordered.  *Only* ref/anon initializers need to be updated.
> 
> Another way to look at this (and maybe name it) would be in terms of named field references vs positional field references; somewhat similar to languages with function parameters that can be initialized either by name or by position at the option of a caller.

That's a crisp description, thanks.

> (I don't see this as blocking, but it would be nice to figure this out for the schema and then make sure the code matches.)

I agree, it would be safe to submit this change as-is. But as a topic of schema design I agree it makes sense to capture.  Added #2128.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-06-14 04:05:02_:

`ref/initializes` (or `ref/init`) seems better than `ref/anon`, to fit in with the other `ref/action` subkinds (rather than `ref/reftype`).

As to the span question, I think that if the field is spelled out (`Pinky: "pink"`), just the field name (`Pinky`) should be linked with a `ref`; otherwise, if it's just the value, the expression should be linked with `ref/init`. Clients that can handle only one reference per character should prioritize `ref/init` links behind other links.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 07:10:41_:

>>! In (earlier comment), zarko wrote:
> `ref/initializes` (or `ref/init`) seems better than `ref/anon`, to fit in with the other `ref/action` subkinds (rather than `ref/reftype`).

Either of those seems fine to me—do you have a preference between `ref/init` and `ref/initializes`?

> As to the span question, I think that if the field is spelled out (`Pinky: "pink"`), just the field name (`Pinky`) should be linked with a `ref`; otherwise, if it's just the value, the expression should be linked with `ref/init`. Clients that can handle only one reference per character should prioritize `ref/init` links behind other links.

I don't think we really want the whole expression linked, since it may wind up being a fairly large recursively-compounded form in some cases. For example,

```
// import "some/path/to/pb" containing
// type Message struct { Field []*pb.Submessage }
// type Submessage struct { P string ; Q int }
//
x := &pb.Message{
   []*pb.Submessage{{
       "foo",
       27,
   },{
       "bar",
       35,
   }},
}
```

Here we'd paint rather a lot of nested anchors, and I'm not sure any extant UI could make much sense of those.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-06-14 17:37:24_:

I prefer `ref/init`.

In your example, where would you put the anchors? I'm not too keen on zero-width spans (and I think we drop them when building serving tables).

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 17:46:26_:

  - Add documentation for ref/init to the schema.
  - Add ref/init to the schema library.
  - Use ref/init for unnamed struct field initializers.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 17:47:55_:

>>! In (earlier comment), zarko wrote:
> I prefer `ref/init`.
> 
> In your example, where would you put the anchors? I'm not too keen on zero-width spans (and I think we drop them when building serving tables).

I would suggest zero-width spans. My rationale is that it provides a location where the initialization takes place, so you can still get a cross-reference at that site—but doesn't force tools to contend with a potentially arbitrary amount of nested structure (often including vertical whitespace, other anchored entities, etc.). And unlike top-level function bodies, the expressions aren't necessarily represented otherwise—we don't paint refs for every typed expression (although we could) because it's hard to use.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-14 17:55:48_:

FWIW, I don't particularly like "ref/init".  AFAICT, these aren't semantically distinct from positional function arguments, should we want to emit references to those at a call site and "ref/init" doesn't make sense in that situation.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 17:57:24_:

  - Add a schema example for ref/init.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 17:59:21_:

>>! In (earlier comment), shahms wrote:
> FWIW, I don't particularly like "ref/init".  AFAICT, these aren't semantically distinct from positional function arguments, should we want to emit references to those at a call site and "ref/init" doesn't make sense in that situation.

What would you suggest instead?

I think we should be careful to distinguish between "could be modeled like function arguments" (which I think is true) and "semantically indistinguishable from function arguments" (which I think is false, or at least not universal). But I don't object to choosing a different name, save that I dislike `ref/anon` because "anonymity" isn't really the important property here.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-14 18:07:01_:

I don't really have a better suggestion.  I guess one could conceptualize the function argument case as "initializing" the argument rather than the field without too many mental gymnastics, but it still feels weird.  It also doesn't answer the (inevitable) question as to why positional fields are "ref/init" while named ones are just "ref". Don't they also initialize the field?

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 19:07:10_:

>>! In (earlier comment), shahms wrote:
> I don't really have a better suggestion.  I guess one could conceptualize the function argument case as "initializing" the argument rather than the field without too many mental gymnastics, but it still feels weird.  It also doesn't answer the (inevitable) question as to why positional fields are "ref/init" while named ones are just "ref". Don't they also initialize the field?

In the key-value version, the key (i.e., the name of the field) is not initializing anything—it's just naming the field being populated, so the name is just a plain reference.

As jdennett pointed out above, there are useful semantic differences between initializing positionally and initializing namewise. You could make the case that the initializing expressions themselves should //always// be references, but I think users would consider that redundant in the case where the key-value initialization is already being done. And: The distinction between positional and key-value is important when refactoring, because order changes affect the former but not the latter, whereas name changes affect the latter but not the former.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-14 19:24:56_:

I agree that the distinction is important, I'm just saying that "ref/init" is confusing (and I guess arguing for "ref/anon").  The fact of "initialization" is not the relevant distinction between the two, it's the positionality/anonymity or lack thereof.  I guess I'd also take "ref/pos" over "ref/init".

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 22:00:34_:

  - Schema: Fix the header name for ref/file.
  - Use the full span for ref/init anchors.
  - Ensure all struct-initializer expressions are annotated.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 22:04:40_:

  - Update the schema example to match #2128.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-06-14 22:05:57_:

Ready for another look.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-06-14 23:57:50_

**Comment by kythe** [comment]
_**fromberger** closed via commit a1b0520558352a17b98ae6ff5760f86f7b38e8d5 at 2017-06-15 00:13:38_
